src/main.cc

```diff
@@ -1443,12 +1434,57 @@ ConfigureCurrentKid(const CommandLine &cmdLine)
     }
 }
 
-static void StartUsingConfig()
+/// Start directing debugs() messages to the configured cache.log.
+/// Until this function is called, all allowed messages go to stderr.
+static void
+ConfigureDebugging()
+{
+    if (opt_no_daemon) {
+        fd_open(0, FD_LOG, "stdin");
+        fd_open(1, FD_LOG, "stdout");
+        fd_open(2, FD_LOG, "stderr");
+    }
+    // we should not create cache.log outside chroot environment, if any
+    // XXX: With Config.chroot_dir set, SMP master process never calls db_init().
+    if (!Config.chroot_dir || Chrooted)
+        _db_init(Debug::cache_log, Debug::debugOptions);
+}
+
+static void
+RunConfigUsers()
 {
     RunRegisteredHere(RegisteredRunner::claimMemoryNeeds);
     RunRegisteredHere(RegisteredRunner::useConfig);
 }
 
+static void
+StartUsingConfig()
+{
+    setMaxFD();
+    fde::Init();
+    const auto skipCwdAdjusting = IamMasterProcess() && InDaemonMode();
+    if (skipCwdAdjusting) {
```

Comment:
Auditing the old vs new startup process I find that this clause can be true when using "-z --foreground" - which will result in the chroot environment being ignored, possibly with wrong paths used or incorrect path errors produced.

Reply:
> this clause can be true when using "-z --foreground" - which will result in the chroot environment being ignored

I think this is not correct. With these arguments, both patched and unpatched Squids spawn a worker and then call mainSetCwd(). chroot is ignored for the parent process, but it was ignored in unpatched Squid too.

Reply:
I also run some tests confirming that mainSetCwd() is called before creating swap directories when running "-z --foreground", i.e, these directories should be created in the correct environment.

Reply:
Testing run-time last night shows that the order is indeed the same in that case, though much later than before (as intended?).

@rousskov , Now that I have the code marked up to display these I'd like to re-run the whole set of permutation checks with backtraces to double-check my notes on the eyeball audit. But if I do not get to that within 5 days feel free to clear this for merge.

Reply:
> if I do not get to that within 5 days feel free to clear this for merge.

Sounds good. And, even if it takes you a bit longer than 5 days, I am sure that we can fix any in-scope bugs your tests expose.

The primary value of this PR, AFAICT, is not the bug it is already fixing but the cleanup of the relevant code that, among other things, exposes several problems that should probably be addressed (separately).